### PR TITLE
Refactor the processAddonLayer function

### PR DESCRIPTION
**Pull request by sbernheim**
@paulcarlton-ww please review and merge.  I added some stuff for testing, but am still working on that part.

### CODE REVIEW

File: pkg/controllers/addons_controller.go [hunk 1]
```diff
@@ -103,13 +90,24 @@ func (r *AddonsLayerReconciler) processAddonLayer(l layers.Layer) error { // nol
...
+	ctx := r.Context
+	applier := r.Applier
 
 	applyIsRequired, err := applier.ApplyIsRequired(ctx, l)
 	if err != nil {
 		// TODO - we might want to add some sort of error handling here
 		return err
 	} else if applyIsRequired {
 		if !l.DependenciesDeployed() {
+			l.SetStatusPending()
```
Comment (paulcarlton-ww) [inline pkg/controllers/addons_controller.go:110]: This is not needed, DependenciesDeployed() sets the status to ApplyPending if one of the layer's dependsOn items are not deployed. Doing it inside DependenciesDeployed() allows us to capture the reason for the the ApplyPending status, i.e. it sets the reason field to say what it is waiting for and why.

File: pkg/internal/layers/layers.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ var rootPath = "/repos"
 type Layer interface {
 	SetStatusK8sVersion()
 	SetStatusApplying()
+	SetStatusPending()
```
Comment (paulcarlton-ww) [inline pkg/internal/layers/layers.go:30]: See above we don't need this

File: pkg/internal/layers/layers.go [hunk 3]
```diff
@@ -197,6 +198,12 @@ func (l *KraanLayer) SetStatusApplying() {
 		kraanv1alpha1.AddonsLayerApplyingReason, kraanv1alpha1.AddonsLayerApplyingMsg)
 }
 
+// SetStatusPending sets the addon layer's status to pending.
```
Comment (paulcarlton-ww) [inline pkg/internal/layers/layers.go:201]: See above we don't need this

**Review by paulcarlton-ww — CHANGES_REQUESTED** [review]
Also you have two TestSingleApply methods

```
Running go lint: /home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply
WARN [runner] Can't run linter goanalysis_metalinter: bodyclose: failed prerequisites: [buildssa@github.com/fidelity/kraan/pkg/internal/apply [github.com/fidelity/kraan/pkg/internal/apply.test]: analysis skipped: errors in package: [/home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:119:6: TestSingleApply redeclared in this block /home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_integration_test.go:234:6:     other declaration of TestSingleApply]] 
WARN [runner] Can't run linter unused: buildir: analysis skipped: errors in package: [/home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:119:6: TestSingleApply redeclared in this block /home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_integration_test.go:234:6:        other declaration of TestSingleApply] 
ERRO Running error: buildir: analysis skipped: errors in package: [/home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:119:6: TestSingleApply redeclared in this block /home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply/applier_integration_test.go:234:6:   other declaration of TestSingleApply] 
/home/a669981/src/github.com/fidelity/kraan/makefile.mk:76: recipe for target '._gometalinter' failed
make[1]: *** [._gometalinter] Error 3
Makefile:122: recipe for target '/home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply' failed
make: *** [/home/a669981/src/github.com/fidelity/kraan/pkg/internal/apply] Error 2
```

File: pkg/controllers/addons_controller.go (hunk 1 above)
Comment (sbernheim) [reply pkg/controllers/addons_controller.go:110]: 👍

File: pkg/internal/layers/layers.go (hunk 2 above)
Comment (sbernheim) [reply pkg/internal/layers/layers.go:30]: 👍

File: pkg/internal/layers/layers.go (hunk 3 above)
Comment (sbernheim) [reply pkg/internal/layers/layers.go:201]: 👍

File: pkg/controllers/addons_controller.go (hunk 1 above)
Comment (sbernheim) [reply pkg/controllers/addons_controller.go:110]: Mainly what I'd wanted to accomplish here was to set `l.updated = true`, which I can now see is triggered by the status updates in the `isOtherDeployed` function in `layers.go` lines 229 and 235.

I'm fine with doing it that way, but I'd consider the status update in a checking function something of an unexpected side-effect to the stated purpose of the function.  It's a little confusing, as evidenced by my impulse to add `SetStatusPending` here.

I'll undo this change regardless - we can always debate over functional programming paradigms over a drink sometime

**Review by paulcarlton-ww — CHANGES_REQUESTED** [review]
test failure
```
=== RUN   TestBasicApply
    applier_test.go:137: NewApplier returned (apply.KubectlLayerApplier) apply.KubectlLayerApplier{client:(*mocks.MockClient)(0xc00042a950), kubectl:(*kubectl.MockKubectl)(0xc00042a930), scheme:(*runtime.Scheme)(0xc00040fc00), logger:testing.TestLogger{T:(*testing.T)(0xc000339b00)}}
    applier_test.go:152: wrong type of argument 0 to Return for *kubectl.MockCommand.WithLogger: *kubectl.MockCommand is not assignable to *kubectl.abstractCommand [/home/pcarlton/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:152]
    panic.go:617: missing call(s) to *kubectl.MockCommand.WithLogger(is equal to {0xc000339b00}) /home/pcarlton/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:152
    panic.go:617: missing call(s) to *kubectl.MockKubectl.Apply(is equal to testdata/apply/single_release) /home/pcarlton/src/github.com/fidelity/kraan/pkg/internal/apply/applier_test.go:151
    panic.go:617: aborting test due to missing call(s)
--- FAIL: TestBasicApply (0.00s)
FAIL
coverage: 2.2% of statements
exit status 1
FAIL    github.com/fidelity/kraan/pkg/internal/apply    0.045s
TEST FAILED
```

**Comment by sbernheim** [comment]
@paulcarlton-ww - the test is not meant to pass at the moment.  Only compile.  I can make it pass, but need to figure out the serializer before I'll really have the test I want.

**Comment by paulcarlton-ww** [comment]
maybe comment it out for now?

**Review by paulcarlton-ww — APPROVED** [review]

**Pull request merged**